### PR TITLE
Improve Performance of Loading Python Plugins During Startup

### LINKED ISSUE #28822: Stop searching for Python plugins when importing mantid.simpleapi
During Framework startup in Python we load `mantid.simpleapi`. As part of the import process we load the Python plugins and this currently requires a [search](https://github.com/mantidproject/mantid/blob/b2f41120f5eeaaa8e24c43b38877b85008d37117/Framework/PythonInterface/mantid/simpleapi.py#L1496) of defined directories and subdirectories for all `.py` files. This is very expensive and made worse on network filesystems or old disks.
We already know the locations of all plugins we bundle with the mantid package and should therefore not need to search for them.

This task is to create a mechanism whereby a "manifest" of known plugins is generated that contains the exact relative locations of all shipped plugins. The simpleapi should then be updated to simply read this file and load the listed plugins.

A [prototype](https://gist.github.com/martyngigg/786a5797a60da1ed145e5253a2e8ab5a#file-simpleapi-py-patch-L426) has been created and it is confirmed to half the import time of `mantid.simpleapi` on a release build of mantid.

### PR DESCRIPTION
**Pull request by cailafinn**
**Description of work.**
Replaces the searching of the file system for a list of python plugins with a CMake generated manifest of paths to those files. This will reduce the time spent accessing the file system during startup when launching mantid on slow drives. 

**To test:**
1. Launch workbench, ideally from a very slow or network drive. 
2. Check that all of the python algorithms have been loaded. 
3. The time taken to launch workbench should also have improved over a version without these changes. 

Fixes #28822 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Properties/Mantid.properties.template [hunk 1]
```diff
@@ -41,6 +41,9 @@ mantidqt.plugins.directory = @QT_PLUGINS_DIR@
 # Where to find python plugins
 python.plugins.directories = @PYTHONPLUGIN_DIRS@
```
Comment (martyngigg) [inline Framework/Properties/Mantid.properties.template:42]: Can we get rid of this key? We shouldn't need it anymore...

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 2]
```diff
@@ -1482,28 +1497,25 @@ def _attach_algorithm_func_as_method(method_name, algorithm_wrapper, algm_object
...
-    for directory in plugin_dirs:
-        try:
-            all_plugins, algs = _plugin_helper.find_plugins(directory)
-            plugin_files.extend(all_plugins)
-            alg_files.extend(algs)
-        except ValueError as exc:
-            logger.warning('Exception encountered during plugin discovery: {0}'.format(str(exc)))
-            continue
+    # Use a cmake generated manifest of all the python algorithms to load them into the api
+    plugins_manifest_path = ConfigService.Instance()["python.plugins.manifest"]
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/simpleapi.py:1501]: There's a couple of cases we need to handle here:
- `plugins_manifest_path` is empty to be to deliberately disable loading
- `plugins_manifest_path` is not a valid path so `open` will raise an `OSError`

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 3]
```diff
@@ -1482,28 +1497,25 @@ def _attach_algorithm_func_as_method(method_name, algorithm_wrapper, algm_object
 # but we need to do this earlier
 setattr(mantid, MODULE_NAME, sys.modules['mantid.{}'.format(MODULE_NAME)])
 try:
-    _plugins_key = 'python.plugins.directories'
-    _user_key = 'user.%s' % _plugins_key
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/simpleapi.py:1486]: We need to keep the old search behaviour for the key `user.python.plugins.directories` as this is how users can write extensions that are found on startup. If the key is not empty can we do the search after we loaded the "shipped" ones from the manifest? Note that the `user.python.plugins.directories` key can contain multiple `;`-separated values to be treated as a list.

File: Framework/PythonInterface/test/python/mantid/kernel/ConfigServiceTest.py [hunk 4]
```diff
@@ -163,8 +163,8 @@ def test_properties_documented(self):
                            'MantidOptions.ReusePlotInstances',
                            'mantidqt.python_interfaces', 'mantidqt.python_interfaces_directory',
                            'paraview.ignore', 'paraview.path', 'paraview.pythonpaths', 'pvplugins.directory',
-                           'python.plugins.directories',
-                       ]
+                           'python.plugins.directories', "python.plugins.manifest"
```
Comment (martyngigg) [inline Framework/PythonInterface/test/python/mantid/kernel/ConfigServiceTest.py:166]: Similarly, can we remove the `python.plugins.directories` entries?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is good. It eliminates a lot of searching on startup for things that are already known upfront. Some comments to address regarding user-provided plugins and error handling.

Can we also update the `concepts/PropertiesFile.rst` document with the remove and added keys?

File: Framework/PythonInterface/mantid/simpleapi.py (hunk 2 above)
Comment (ConorMFinn) [reply Framework/PythonInterface/mantid/simpleapi.py:1501]: If the manifest is empty should we be launching anyway or fail out with an appropriate error?

Comment (martyngigg) [reply Framework/PythonInterface/mantid/simpleapi.py:1501]: We can still launch. If a user wants to disable the python algorithms then as long as they don't try to call one it should be fine.

Comment (martyngigg) [reply Framework/PythonInterface/mantid/simpleapi.py:1501]: It would be the same as blanking the `python.plugins.directories` currently.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is really close now. Just a few more minor comments.

**Comment by martyngigg** [comment]
@quantumsteve @peterfpeterson This might help with the slow loading issues you were having on your network file systems?

**Review by martyngigg — DISMISSED** [review]
This looks good to me now.

**Comment by gemmaguest** [comment]
@ConorMFinn This has conflicts

**Comment by gemmaguest** [comment]
@martyngigg Could you re-approve this? The doxygen build seems to have passed even though it's showing as pending!

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@gemmaguest Done. Sometimes the notifications seemed to get lost on the way back for some reason.

**Pull request merged**